litestar/_signature/models/msgspec_signature_model.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
+
+import re
+from typing import TYPE_CHECKING, Any, Optional, Union
+
+from msgspec import NODEFAULT, Meta, Struct, ValidationError, convert, defstruct
+from msgspec.structs import asdict
+from pydantic import ValidationError as PydanticValidationError
+from typing_extensions import Annotated
+
+from litestar._signature.field import SignatureField
+from litestar.params import DependencyKwarg
+from litestar.serialization import dec_hook
+from litestar.types.empty import Empty
+from litestar.utils import make_non_optional_union
+from litestar.utils.dataclass import simple_asdict
+from litestar.utils.typing import unwrap_union
+
+from .base import SignatureModel
+
+if TYPE_CHECKING:
+    from litestar.connection import ASGIConnection
+    from litestar.utils.signature import ParsedSignature
+
+    from .base import ErrorMessage
+
+__all__ = ("MsgspecSignatureModel",)
+
+MSGSPEC_CONSTRAINT_FIELDS = (
+    "gt",
+    "ge",
+    "lt",
+    "le",
+    "multiple_of",
+    "pattern",
+    "min_length",
+    "max_length",
+)
+
+ERR_RE = re.compile(r"`\$\.(.+)`$")
+
+
+class MsgspecSignatureModel(SignatureModel, Struct):
+    """Model that represents a function signature that uses a msgspec specific type or types."""
+
+    @classmethod
+    def parse_values_from_connection_kwargs(cls, connection: ASGIConnection, **kwargs: Any) -> dict[str, Any]:
+        """Extract values from the connection instance and return a dict of parsed values.
+
+        Args:
+            connection: The ASGI connection instance.
+            **kwargs: A dictionary of kwargs.
+
+        Raises:
+            ValidationException: If validation failed.
+            InternalServerException: If another exception has been raised.
+
+        Returns:
+            A dictionary of parsed values
+        """
+        messages: list[ErrorMessage] = []
+        try:
+            return convert(kwargs, cls, strict=False, dec_hook=dec_hook).to_dict()
+        except PydanticValidationError as e:
+            for exc in e.errors():
+                keys = [str(loc) for loc in exc["loc"]]
+                message = super()._build_error_message(keys=keys, exc_msg=exc["msg"], connection=connection)
+                messages.append(message)
+            raise cls._create_exception(messages=messages, connection=connection) from e
+        except ValidationError as e:
+            match = ERR_RE.search(str(e))
+            keys = [str(match.group(1)) if match else "n/a"]
```

Comment:
Unlike the others, msgspec will only report one error: this is likely the biggest point of reckoning for the PR.

@jcrist is this something that can be explored on your side of the fence?

Using either pydantic or cattrs, we can report all validation errors, not just the first encountered. For example, if a client submits an invalid header value, and an invalid request body we can report both of those failures in the error response.

One possible way this could be handled on our side would be to have separate Structs to validate each of the data sources we have: body, query, cookies and headers (if DTOs are in use, we already do make a separate `convert()` call for body). This would allow us to report the _first_ error from each source, but would also require up to 4 separate calls to `convert()` for a request, which might be a hit to performance (although that would have to be bench marked before we lose sleep over it).

Reply:
Actually, this will also affect our support for attrs (although we haven't had that in a proper release yet). The current attrs support comes from decoding with cattrs, which gives us all validation errors for a model as a validation group.

Take this example:

```py
import attrs
import cattrs
import msgspec
from exceptiongroup import ExceptionGroup

@attrs.define
class Foo:
    a: int
    b: str

try:
    cattrs.global_converter.structure({"data": {"one": 1, "b": 2}}, Foo)
except ExceptionGroup as e:
    print(str(e))
    print(e.exceptions)

try:
    msgspec.convert({"data": {"a": "one", "b": 2}}, type=Foo)
except Exception as e:
    print(str(e))
```

From cattrs we get an exception group and can access underlying exceptions for each sub-error:

> While structuring Foo (2 sub-exceptions)
(KeyError('a'), KeyError('b'))

From msgspec we only get the the one error:

> Object missing required field `a`

From my understanding, this cannot be solved with a dec hook either, as given msgspec's native support for attrs the dec hook would never be called.

Reply:
> @jcrist is this something that can be explored on your side of the fence?

No, this is not something I intend on supporting. For prior art, most typed JSON serialization libraries (across language ecosystems) don't support raising multiple errors. Neither golang's `json` nor rust's `serde` support this, and people generally love those tools.

When parsing JSON into a specified object type, there are multiple ways an error can occur:
- The JSON is invalid
- The JSON is valid, but the JSON type is incorrect (expected a `number`, got an `array`)
- The JSON type is correct, but the value is invalid (e.g. unknown Enum value, int out of range, invalid datetime str etc...)
- A JSON object is missing a field (or in certain situations has an extra field)

When parsing and validating, a failure in one location could be one error, or several. There's no clearly defined behavior for what should be expected. `pydantic` might group a few errors together and say "these are the things that are invalid". `cattrs` might find a different set of errors. But there's no clear definition of what "all the errors" means. Say for example you're parsing the value `"2022-03-"` into a type `datetime | date | None`. This string is not a valid datetime or date, and it isn't None. Is that one, two, or three errors?

Further, what if you receive a list of 1000 strings as an input for a field, but expected a list of ints. Is that 1000 errors (one for each item), or one for the list?

Maintaining a running structure of errors like this can get expensive, and make the failure mode in parsing significantly more expensive than the success mode. I'm skeptical of its general utility, and think if providing errors for multiple fields is important, you'd be better served by doing that validation yourself _after_ parsing, so you can ensure the kinds of user-facing errors you wish to provide are uniformly handled, rather than whatever errors the parsing framework happens to collect.

--- 

> One possible way this could be handled on our side would be to have separate Structs to validate each of the data sources we have: body, query, cookies and headers (if DTOs are in use, we already do make a separate convert() call for body).

I think this is a good idea, and is what I recommended y'all do several months ago: https://github.com/jcrist/msgspec/issues/301#issuecomment-1434842236.

Besides making it easier to group errors to include in a response, it makes it easier to make parsing more specific to the provided types. AFAICT right now you're combining all data sources (headers, body, ...) into a big dict, then passing that to an uberfunction to handle conversion into the types. This has a few downsides:

- It prevents optimizing the JSON decoding by parsing directly into the type using `msgspec.json.decode`, since you're decoding into a dict first then using `msgspec.convert` to convert. This has an unnecessary performance cost, and is missing out on the main reason I wrote msgspec in the first place.
- It prevents tailoring the decoding settings to the protocol at hand. When parsing data from untyped sources like headers/query parameters (which are all strings) you'll want very flexible coercion rules; you'll likely want to set `strict=False` (https://jcristharif.com/msgspec/usage.html#strict-vs-lax-mode). But when parsing something more structured like JSON those lax coercion rules can mask client-side issues, you likely want `strict=True` here. Pushing all parsing and validation into a single call prevents tailing the specifics of the parsing to the input data source.
- With dependency injection, some of the inputs to a handler likely don't need validation, since they're coming from within the litestar framework itself, not user data. Grouping these inputs into the same input dict and passing them to `msgspec.convert` is doing unnecessary work (though it may be pretty negligible in cost).

4 calls to `msgspec.convert` instead of a single call should have negligible performance impact, and if it lets you call a more specific parsing function for certain inputs (e.g. `msgspec.json.decode` for parsing json bodies) then this should have a net performance win.

> From my understanding, this cannot be solved with a dec hook either, as given msgspec's native support for attrs the dec hook would never be called.

This is correct, if you're using msgspec to parse a type containing an `attrs` type, msgspec will always handle this parsing itself rather than falling back to a `dec_hook`. The system was not designed to coexist with, but rather to supplant the need for `cattrs`.

If you separate out the input sources though like I recommended above, you could do something like:

- if the body type is an attrs object, use cattrs to parse it from json
- if the body type is a pydantic object, use pydantic to parse it from json
- otherwise use msgspec to parse it from json

Reply:
Thanks for your detailed response @jcrist!

> No, this is not something I intend on supporting.

Understood, I can't argue with any of your rationale, and knowing this is off the table is just as helpful with moving the PR forward as it being on the table.

> I think this is a good idea, and is what I recommended y'all do several months ago

:grimacing: Sorry, I wasn't aware of that prior discussion, but good to know we are on the same page.

> 4 calls to msgspec.convert instead of a single call should have negligible performance impact

And, if we are happy enough to group the sources that we pre-decode together (query, headers, cookies), then this really only needs to be 2.

> If you separate out the input sources though like I recommended above, you could do something like:
>
>   - if the body type is an attrs object, use cattrs to parse it from json
>   - if the body type is a pydantic object, use pydantic to parse it from json
>   - otherwise use msgspec to parse it from json

I agree, this makes the most sense. Also then, we wouldn't need to worry about the exception chaining in order to access those pydantic exceptions, as I mentioned in another comment.